src/tools/compiletest/src/runtest.rs

```diff
@@ -3790,11 +3813,37 @@ impl<'test> TestCx<'test> {
         }
     }
 
-    fn compare_output(&self, kind: &str, actual: &str, expected: &str) -> usize {
+    fn compare_output(
+        &self,
+        kind: &str,
+        actual: &str,
+        expected: &str,
+        compare_output_by_lines: bool,
+    ) -> usize {
         if actual == expected {
             return 0;
         }
 
+        let tmp;
+        let (expected, actual): (&str, &str) = if compare_output_by_lines {
+            let expected_lines: Vec<_> = expected.lines().collect();
+            let mut used = expected_lines.clone();
+            used.retain(|line| actual.lines().any(|l| l == *line));
+            // check if `expected` contains a subset of the lines of `actual`
+            if used.len() == expected_lines.len() && (expected.is_empty() == actual.is_empty()) {
```

Comment:
I think you can get this with putting both sets of lines into HashSet and then calling https://doc.rust-lang.org/nightly/std/collections/hash_set/struct.HashSet.html#method.is_subset, while also avoiding both the - if I'm getting the math right - O(n^3) (n^2 for each set and another n for the copies as we remove lines, I think, in the worst case) retain loop.

Reply:
Isn't this already n^2? For each expected line we iterate through the actual output once. That aside, using a `HashSet` doesn't work as we join over the vecs in line 3840 which is order dependent. Unless `IndexSet` is used which requires adding it as a dependency to `compiletest`.